include/maliput/math/kd_tree.h

```diff
@@ -264,5 +344,148 @@ class KDTree {
+  void InitializeRegions() {
+    this->root_->set_region(std::make_unique<AxisAlignedBox>(
+      Vector3{-std::numeric_limits<double>::infinity(), -std::numeric_limits<double>::infinity(), -std::numeric_limits<double>::infinity()},
+      Vector3{std::numeric_limits<double>::infinity(), std::numeric_limits<double>::infinity(), std::numeric_limits<double>::infinity()}));
+    if(this->root_->get_left() != nullptr)details::Initialize3dRegions(true, this->root_->get_left());
+    if(this->root_->get_right() != nullptr)details::Initialize3dRegions(false, this->root_->get_right());
+    initialized_regions = true;
+  }
+
+ private:
+  using KdTreeBaseNode = typename KDTreeBase<KDTree3D<Coordinate, Distance, NodeCmp>, Coordinate, 3, AxisAlignedBox, Distance, NodeCmp>::Node;
+
+  // Performs a range search query recursively.
+  // @param node The node to be searched.
+  // @param region The search region.
+  // @param result The result of the search.
+  void RangeSearch(KdTreeBaseNode* node, const AxisAlignedBox& region, std::deque<const Coordinate*>& result) const {
+    if(!initialized_regions) {
+      maliput::log()->error("Regions weren't initialized. For using KDTree3D::RangeSearch method, you need to call KDTree3D::InitializeRegions() first.");
+      return;
+    }
+    if (node == nullptr) return;
+    // If the node is a leaf, check if its coordinate is in the region.
+    if (node->get_left() == nullptr && node->get_right() == nullptr) {
+      const auto& coordinate = node->get_coordinate();
+      if (region.Contains({coordinate[0], coordinate[1], coordinate[2]})) {
+        result.push_back(&coordinate);
+      }
+      return;
+    }
+    // If the region is fully contained in the node, report the node and its children.
+    const AxisAlignedBox node_region = dynamic_cast<const AxisAlignedBox&>(node->get_region());
+    const auto overlapping_type = region.Overlaps(node_region);
+
+    if(overlapping_type == OverlappingType::kContained) {
+      ReportPointsInSubTree(node, result);
+      return;
+    }
+    if(overlapping_type == OverlappingType::kIntersected) {
+      const auto& coordinate = node->get_coordinate();
+      if (region.Contains({coordinate[0], coordinate[1], coordinate[2]})) {
+        result.push_back(&coordinate);
+      }
+      RangeSearch(node->get_left(), region, result);
+      RangeSearch(node->get_right(), region, result);
+      return;
+    }
+    if(overlapping_type == OverlappingType::kDisjointed) {
+      return;
+    }
+  }
+
+  // Reports the points in the subtree rooted at @p node.
+  // @param node The node to be searched.
+  // @param result The result of the search.
+  void ReportPointsInSubTree(KdTreeBaseNode* node, std::deque<const Coordinate*>& result) const {
+    // TODO(francocipollone): Each node could have a list of references to points located within the region so we don't
+    //                        have to iterate all the nodes for getting the nodes that are contained.
```

Comment:
We discussed F2F yesterday about this improvement. I guess it could be worth trying and see how it behaves.

Reply:
Agreed. Let's push it to a second order optimization. I think it won't change that much.